UnchainedLauncherCore/src/API/ServerBrowser.cs

```diff
@@ -0,0 +1,248 @@
+    );
+    [AddINotifyPropertyChangedInterface]
+    public record C2ServerInfo {
+        public bool PasswordProtected { get; set; } = false;
+        public string Name { get; set; } = "";
+        public string Description { get; set; } = "";
+        public Ports Ports { get; set; } = new(7777, 3075, 7071);
+        // TODO: The selection of the Mod datatype here is potentially
+        // incorrect. Change it to whatever is easier and works with the backend
+        public Mod[] Mods { get; set; } = Array.Empty<Mod>(); 
+    };
+    // TODO: This part of the API should be stabilized
+    // as-is, there are numerous different kinds of "server" objects depending
+    // on where they're coming from/where they're going
+    public record ServerInfo : C2ServerInfo{
+        public String CurrentMap { get; set; } = "";
+        public int PlayerCount { get; set; } = 0;
+        public int MaxPlayers { get; set; } = 0;
+        public ServerInfo() { }
+        public ServerInfo(C2ServerInfo info, A2S_INFO a2s) : base(info) {
+            Update(a2s);
+        }
+        public bool Update(A2S_INFO a2s)
+        {
+            bool wasChanged = (
+                MaxPlayers != a2s.MaxPlayers
+                || PlayerCount != a2s.Players
+                || CurrentMap != a2s.Map);
+            MaxPlayers = a2s.MaxPlayers;
+            PlayerCount = a2s.Players;
+            CurrentMap = a2s.Map;
+            return wasChanged;
+        }
+    }
+
+    public record UniqueServerInfo : ServerInfo {
+        public String UniqueId { get; set; } = "";
+        public double LastHeartbeat { get; set; } = 0;
+    };
+    
+    public record RegisterServerRequest : ServerInfo
+    {
+        public String LocalIpAddress { get; set; }
+        public RegisterServerRequest(ServerInfo info, string localIpAddress) : base(info)
+        {
+            this.LocalIpAddress = localIpAddress;
+        }
+    }
+
+    public record ResponseServer : UniqueServerInfo
+    {
+        public String LocalIpAddress { get; set; } = "";
+        public String IpAddress { get; set; } = "";
+        public ResponseServer() { }
+        public ResponseServer(UniqueServerInfo info, string localIpAddress, string ipAddress) : base(info)
+        {
+            this.LocalIpAddress=localIpAddress;
+            this.IpAddress=ipAddress;
+        }
+    }
+    public record RegisterServerResponse(
+        double RefreshBefore,
+        String Key,
+        ResponseServer Server
+    );
+    public record UpdateServerRequest(
+        int PlayerCount,
+        int MaxPlayers,
+        String CurrentMap
+    );
+
+    public record UpdateServerResponse(double RefreshBefore, ResponseServer Server);
+
+    record GetServersResponse(
+        UniqueServerInfo[] Servers  
+    );
+    //URI is expected to have the /api/v1/ stuff. These functions only append the immediately relevant paths
+    public static class ServerBrowser
+    {
+        static readonly HttpClient httpc = new();
```

Comment:
`HttpClient` should be a parameter provided to the constructor of `ServerBrowser`. We can have some kind of memoized `ServerBrowser.Default` static method which creates and caches a server browser with an http client configured with some sane defaults

Reply:
What is the main benefit to doing it this way? If we make a memoized httpclient creator, then we make a fairly static HTTP client anyway. I made these static functions so that the implementation hides the creation of a bunch of objects that'd need to hang around between calls. This way, they can be called as if they were native functions that can time out. If we require instances of this class to be made, then the user has to carry and pass instances of it around, or just repeatedly make instances of it on-demand.

HttpClients are very thread-safe and meant to be aggressively re-used, so I think having the lib handle that re-use rather than obligating the users to do it is the best option. This way, multiple different RegisteredServer's are using the same HttpClient as well. The only issue is that this means HttpClient state can't be set on a per-request basis, which we don't currently do. (changes to how the timeout works will probably change this, but I question how important it is to have fine-grained control over the timeouts.)